Structure the UI

crazyscientist opened the pull request:
* Added components to encourage common layout
* Moved schema list into navigation bar
* Return actual schema data from `API.getSchemas`

crazyscientist committed:
Structure the UI

* Added components to encourage common layout
* Moved schema list into navigation bar
* Return actual schema data from `API.getSchemas`

commanderprice commented on frontend/src/main.js:
```diff
@@ -3,4 +3,5 @@ import App from './App.vue'
 import { router } from './router'
 
 const app = createApp(App).use(router);
+app.config.unwrapInjectedRef = true;
```
What's this for?🤔

crazyscientist replied:
When I added the `provide`/`inject` elements I struggled a bit, because it did not work as expected. It turned out that the injected object is wrapped and needs explicit unwrapping in the current version of Vue. But it is planned to automatically unwrap it in a future release. 

This option enables this behavior in the current version.

commanderprice commented on frontend/src/components/layout/BaseLayout.vue:
```diff
@@ -0,0 +1,20 @@
+<template>
```
I looked at this component usages, but I'm not sure that I got it's purpose: is it just for that top part with breadcrumbs?

crazyscientist replied:
Basically yes. My intent was to create a generic layout component, that we can use in all router views :slightly_smiling_face:

commanderprice commented on frontend/src/components/SchemaList.vue:
```diff
@@ -1,23 +1,86 @@
+        }
+      }
+    },
+    selectSchema(event) {
```
I guess it's possible to pass schema id to onclick func and have `schemas` as mapping `{id: schema}` to just get schema by id instead of iterating over all schemas

crazyscientist replied:
When I started to mess around with the UI, I thought this might be a good approach, but now I'm not that sure anymore. I will need to change this part a little more...

crazyscientist replied:
Found a different approach: Schema selection is now done solely via the URL :grin:

commanderprice commented on frontend/src/components/SchemaList.vue:
```diff
@@ -1,23 +1,86 @@
+      return "Schema";
+    }
+  },
+  mounted: function () {
```
Probably we should have `created` here instead of `mounted`🤔

crazyscientist replied:
That should also work, yes :slightly_smiling_face:

crazyscientist replied:
Changed it.

commanderprice commented on frontend/src/components/SchemaList.vue:
```diff
@@ -1,23 +1,86 @@
-      {{ schema.name }}
-    </li>
-  </ul>
+  <li class="nav-item dropdown">
```
In general I like this option more than my initial when schemas list was sitting all the time to the left of entities list taking up space, but:
1. I guess we need ordering of schemas by name here (currently it displays them just as they were received)
2. Maybe we need to deal somehow with a big list of schemas. For example, if we have a list of 100 schemas it will just open this dropdown with all schemas and with vertical scroll of several display heights. But maybe that's okay🤔

crazyscientist replied:
I think/hope that the number of schemas should be relatively small.

On the other hand we will also need some feature to find/access deleted schemas. I was hoping to address all of that in a second PR...

commanderprice commented on frontend/src/api.js:
```diff
@@ -13,7 +28,14 @@ class API {
         params.set('all', all);
         params.set('deleted_only', deletedOnly);
-        return fetch(`${this.base}/schemas?${params.toString()}`);
+        const response = await fetch(`${this.base}/schemas?${params.toString()}`);
+        try {
+            this._is_response_ok(response);
+            return response.json()
+        }
+        catch (e) {
+            // TODO: Do something else
+        }
```
1. I'm okay with returning json directly from function, but why is it just here?🤔
2. I'm not sure that we should catch exceptions at API functions, or at least our default behavior should be to reraise them so that in views we could decide how to react on specific error

crazyscientist replied:
> but why is it just here?

I did not get through all the views and code, I only got around to change this one. If you like the general idea, that it great to hear and I will continue with refactoring the other methods.

> I'm not sure that we should catch exceptions at API functions

I'm afraid that a lot of exception handling in the views would look very similar. So I was trying to centralize exception handling. but we could extend it by passing custom error handler functions to the API method calls :thinking:

commanderprice replied:
> passing custom error handler functions

This could possibly work, I guess. The idea is when getting exception, in most cases (if not all) user should see that something went wrong and preferably what exactly which is something that has to be defined in views. If vuejs functions keep correct `this` context when passed to other (non-vue) functions then great, we can try passing exception handlers from views to API calls

crazyscientist replied:
Error and success messages are now pushed onto a queue from which they are displayed in the UI.

commanderprice commented on frontend/src/components/EntityList.vue:
```diff
@@ -1,105 +1,106 @@
 <template>
-  <div class="row">
-    <div class="col-lg-2">
+  <BaseLayout>
+    <template v-slot:additional_breadcrumbs>
+      <li class="breadcrumb-item">{{ schemaName }}</li>
+      <li class="breadcrumb-item active">Entities</li>
+    </template>
+    <template v-slot:actions>
```
This breadcrumb component has a little bit unstable behavior across pages; I don't know exactly which is intended and which is not, so here are my observations

Action   |  What is shown in breadcrumbs
---------|-----------------------------------
Open start page  | `home / --- / Entities`
Choose schema |  `home / Schema name / Entities`
Go to entity detail | None
Go back, open schema details | `home / Schema name / Details`
Click edit | `home / Schema name / Details`
Go back to details | `home / Schema name / Details` <br> `home / Schema name / Details`
Go back to entity list |  `home / Schema name / Details`  <br> `home / Schema name / Details` <br>`home / Schema name / Entities`
Refresh, choose schema | `home / Schema name / Entities`
Click edit | None
Go back, click create entity | None
Go back, open entity details, click edit | None
Go back to list, click create new schema | None
Go back, open schema details |  `home / Schema name / Details`
Click create new schema | `home / Schema name / Details` (and edit button leading to schema `undefined`)
UPD | UPD
Open entity list, choose schema |  `home / Schema name / Entities`
Open schema details |  `home / Schema name / Details`
Click edit | `home / Schema name / Details`
Go to entity list via link in navbar | `home / Schema name / Details` <br>`home / Schema name / Entities`

By _go back_ here I mean _go back in browser history_

crazyscientist replied:
That was indeed not so nice... But as I continued to work my way through the components, this unwanted behavior became visible at less places.

I will recheck, once I'm finished digging through the UI.

commanderprice replied:
I love self-eliminating bugs😄

crazyscientist committed:
Major refactoring of UI

* Use named routes
* Merged similar components into one
* Translated imports to absolute paths
* Show schema list on landing page
* Added new base component for inputs

commanderprice commented on frontend/src/components/EntityList.vue:
```diff
@@ -109,59 +61,56 @@ export default {
-      const json = await response.json();
-      this.entities = json.entities;
-      this.totalEntities = json.total;
-      this.operators = json.meta.filter_fields.operators;
-      this.filterableFields = json.meta.filter_fields.fields;
```
You forgot to move this part to `getEntities()` -- without this we don't have list of filterable fields and operators for them

crazyscientist replied:
Oops, we definitely want to keep that. :+1:

commanderprice commented on frontend/src/components/EntityList.vue:
```diff
@@ -171,14 +120,14 @@ export default {
     };
   },
+  mounted() {
+    if (this.schema !== undefined && this.schema !== null) {
```
Can't this be simplified to `if (this.schema)`?🤔

crazyscientist replied:
Yes, we can simplify it to `if (!this.schema)` :wink:

commanderprice commented on frontend/src/components/EntityList.vue:
```diff
@@ -1,104 +1,56 @@
+    <div class="col-2">
+      <label for="entitiesLimit"><small>Entities per page</small></label>
+    </div>
+    <div class="col-1">
+      <select v-model="entitiesPerPage" id="entitiesLimit" class="form-select form-select-sm">
+        <option>10</option>
+        <option>30</option>
+        <option>50</option>
+      </select>
+    </div>
+    <div class="col-9 text-end">
```
I'd add something like `col-sm-2` for first div and `col-sm-8` for second one because at ~982px of width you can't really see what's selected

crazyscientist replied:
Right. I'd go even further and say that `col`s don't fit well here. I've replaced it with a `d-flex` :grin:

commanderprice commented on frontend/src/components/EntityList.vue:
```diff
@@ -109,59 +61,56 @@ export default {
+      this.getEntities({resetPage: true});
     },
-    async reorder({ orderBy, ascending } = {}) {
+    async reorder({orderBy, ascending} = {}) {
       this.orderBy = orderBy;
       this.ascending = ascending;
       this.getEntities();
```
Currently if you order by some field (not name) and switch to other schema, it be in constantly loading state. Quick research showed that placing 
```
this.orderBy = 'name';
this.ascending = true;
```
in `schema()` watch kind of helps but it looks like `EntityListTable` still keeps its state with initially ordered field

commanderprice replied:
Okay, I found it. In `updated()` of `EntityListTable` we are checking `this.schemaSlug`, but since you are now passing whole schema to component the check should be this instead:
```
updated() {
    if (this.previousSchema !== this.schema.slug) {
      this.previousSchema = this.schema.slug;
      this.ascending = true;
      this.orderBy = 'name';
    }
```

crazyscientist replied:
Thanks for noticing. Fixed.

commanderprice commented:
How to create new entity now?🤔

commanderprice commented on frontend/src/components/SchemaList.vue:
```diff
@@ -1,23 +1,77 @@
-      :class="selectedSchema === schema ? 'active' : ''"
-    >
-      {{ schema.name }}
+  <template v-if="asDropdown">
+    <li class="nav-item dropdown">
+      <a class="nav-link dropdown-toggle" href="#" id="nav-schema-dropdown" role="button"
```
When switching between schemas with dropdown attribute list in schema edit form doesn't change; when you switch with `Schema selection` it's okay

crazyscientist replied:
Fixed. By converting the dropdown items to links the behavior is identical.

crazyscientist replied:
OK, this still seems to be an issue :slightly_frowning_face:

crazyscientist replied:
OK, now it's fixed.

commanderprice commented:
Schema create form looks drunk🤪 Also here it doesn't have list of available schemas for FK fields

![image](https://user-images.githubusercontent.com/36228624/146144265-1a1bf6e0-1a8d-4637-9f40-fed6aa9eda34.png)

commanderprice commented on frontend/src/App.vue:
```diff
@@ -1,30 +1,96 @@
+      if (schemaSlug === undefined || schemaSlug === null) {
+        return null;
+      }
+      try {
+        // First, try to reuse data in storage
+        this.activeSchema = this.availableSchemas[schemaSlug];
+      } catch (e) {
+        true;
+      }
+      this.getSchemaFromApi(schemaSlug);
```
> try to reuse data in storage

Not sure that it's a good idea, because schema could have changed🤔 Although they are not supposed to be changed frequently.

But even so, it doesn't look that `this.activeSchema = this.availableSchemas[schemaSlug];` has any effect since after all it gets to `this.getSchemaFromApi(schemaSlug);` which sets schema on its own🙃

crazyscientist replied:
Good point. 

If I remember correctly, here we only care about the minimal schema information (name and slug to be exact). It might be safe to assume that these do not change often, as you already pointed out.

So, for the time being, I'd suggest to actually implement what I intended (get element from already fetched list first. If not possible, fetch from API).

commanderprice replied:
> I'd suggest to actually implement what I intended

okay👌

commanderprice commented on frontend/src/App.vue:
```diff
@@ -1,30 +1,96 @@
+    },
+    getSchemaFromRoute() {
+      let schemaSlug = this.$route.params.schemaSlug || null;
+      if (schemaSlug === undefined || schemaSlug === null) {
```
Can this be simplified to `if (!schemaSlug)`?

crazyscientist replied:
Yes :+1:

commanderprice commented on frontend/src/App.vue:
```diff
@@ -1,30 +1,96 @@
-  components: {
-
-  }
+  metaInfo: {
+    title: 'SUSE aimaas'
+  },
```
Is this used somehow/by something?🤔

crazyscientist replied:
Oops, I was trying to set the page title, but did not succeed. I will remove it for the time being.

commanderprice commented on frontend/src/components/SchemaEdit.vue:
```diff
@@ -1,131 +1,105 @@
+                    :availableFieldNames="availableFieldNames"/>
+    <button @click="sendData" type="button" class="mt-3 mb-3 btn btn-outline-primary">
+      <i class='eos-icons'>save</i>
+      Update schema
```
Not technically an error but there's no visual response when updating schema so you are not sure if anything has really changed

crazyscientist replied:
The idea also occurred to me. I guess I will add something :+1:

crazyscientist replied:
Now, the button is disabled by default. Only after a change, the button becomes enabled.

Is this what you had in mind?

commanderprice replied:
Not exactly. This one is useful too, because it let's you know that you've actually changed something; but when you change something and click on update button you don't see any effect. It could be some popup telling that schema was updated or forced page reload to show updated schema — anything that would tell user that click on update button changed something

crazyscientist replied:
Thanks for the clarification.

I was hoping to keep this PR free of new features, i.e. that I can only focus on the structure. Would you like to get everything in one PR? Or can we take care of this later in a separate PR?

commanderprice replied:
> Or can we take care of this later in a separate PR?

Sure👍

crazyscientist replied:
Oops, I have added that already to this PR.... :innocent:

commanderprice commented:
Entities can't be edited now. Judging from what I see in console, when it tries to get entity details it loads `undefined` entity from server

crazyscientist commented:
> Entities can't be edited now. Judging from what I see in console, when it tries to get entity details it loads undefined entity from server

Yes, sadly I did not manage to get through everything yet. I hope I will get it done soon :crossed_fingers:

crazyscientist committed:
Applied suggestions

crazyscientist committed:
Adjusted search panel

crazyscientist committed:
Finished work on `SchemaEdit`

crazyscientist committed:
Rolled back unintentional backend change

crazyscientist committed:
Fixed ordering issue when switching schemas

crazyscientist committed:
Fixed schema creation (and use forms from schema edit ;) )

commanderprice commented on frontend/src/App.vue:
```diff
@@ -1,30 +1,96 @@
+        return null;
+      }
+      try {
+        // First, try to reuse data in storage
```
I was  wondering, won't it make sense to put `this.getSchemaFromApi(schemaSlug);` inside of catch?🤔

crazyscientist replied:
That sounds feasible at the moment. :+1: 
But it will depend on how we decide to handle errors in the API object.

commanderprice commented on frontend/src/components/SchemaEdit.vue:
```diff
@@ -141,7 +115,7 @@ export default {
       toSend.add_attributes = add;
```
I think that this needs to be 
```
const toSend = {
        name: this.details.name,
        slug: this.details.slug,
        reviewable: this.details.reviewable,
      };
```
or something like that, because currently updating these fields changes nothing

crazyscientist replied:
I will need to check. I also noticed, that deleting attributes had no effect (but I also don't remember whether my schema required reviews :thinking: )

commanderprice replied:
> that deleting attributes had no effect...whether my schema required reviews

Nah, I think it was not the case😅 I somehow managed to miss this part in traceability for schema where it applies update request. So, although there is needed data to perform deletion of attribute, program just skips it🙄 It doesn't make much sense to fix this at current traceability state, because it will be rewritten anyway. But if you need this to work,
in `traceability/schema.py` in change `apply_schema_update_request()` to this
```python
def apply_schema_update_request(db: Session, change_id: int, reviewed_by: User, comment: Optional[str] = None) -> Schema:
    change = db.execute(
        select(Change)
        .where(Change.id == change_id)
        .where(Change.change_object == ChangeObject.SCHEMA)
        .where(Change.change_type == ChangeType.UPDATE)
    ).scalar()
    if change is None:
        raise MissingSchemaUpdateRequestException(obj_id=change_id)
    schema_update: SchemaUpdate = db.execute(select(SchemaUpdate).where(SchemaUpdate.change_id == change.id)).scalar()
    attrs_update: List[AttributeUpdate] = db.execute(select(AttributeUpdate).where(AttributeUpdate.change_id == change.id)).scalars().all()
    attrs_create: List[AttributeCreate] = db.execute(select(AttributeCreate).where(AttributeCreate.change_id == change.id)).scalars().all()
    attrs_delete: List[AttributeDelete] = db.execute(select(AttributeDelete).where(AttributeDelete.change_id == change.id)).scalars().all()
    data = SchemaUpdateSchema(
        name=schema_update.new_name,
        slug=schema_update.new_slug,
        reviewable=schema_update.new_reviewable,
        update_attributes=[
            AttrDefUpdateSchema(
                name=i.attribute.name,
                new_name=i.new_name,
                required=i.required,
                unique=i.unique,
                list=i.list,
                key=i.key,
                description=i.description,
                bind_to_schema=i.bind_to_schema
            )
            for i in attrs_update
        ],
        add_attributes=[
            AttrDefSchema(
                name=i.name,
                type=i.type.name,
                required=i.required,
                unique=i.unique,
                list=i.list,
                key=i.key,
                description=i.description,
                bind_to_schema=i.bind_to_schema
            )
            for i in attrs_create
        ],
        delete_attributes=[i.attribute.name for i in attrs_delete]
    )
    change.reviewed_at = datetime.utcnow()
    change.reviewed_by = reviewed_by
    change.status = ChangeStatus.APPROVED
    change.comment = comment
    s = update_schema(db=db, id_or_slug=schema_update.schema_id, data=data, commit=False)
    db.commit()
    return s
```

crazyscientist replied:
OK, if you are confident that I did not break it on the frontend, it's fine for me at the moment :slightly_smiling_face:

commanderprice replied:
Well, at least the last time I checked the problem was only on backend side😏

commanderprice commented on frontend/src/components/SchemaEdit.vue:
```diff
@@ -151,24 +160,26 @@ export default {
-      if (response.status == 200) {
-        this.$router.push(`/${this.slug}`);
+      if (response.status === 200) {
+        this.$router.push({name: 'schema-view', params: {schemaSlug: this.schema.slug}});
```
Here you are redirecting to slug of schema which was initially there, but if we change slug we'll get 404 error. And you can't just use `this.details.slug` here instead — if schema is reviewable, then this update (possibly) won't be applied immediately so there won't be a schema in database that has slug contained in `this.details.slug`. I guess to handle that it's possible to analyze response content or even have different 2xx for cases when something was updated immediately and when there was created change request.

But that also will require some changes in backend, because currently for some reason it just creates request and applies it immediately🙄

crazyscientist replied:
Good point. We could just not do redirects and show nice success/error messages instead? :thinking:

commanderprice replied:
That's an option too, although I think that redirects still make sense🙃

commanderprice commented on frontend/src/components/SchemaCreate.vue:
```diff
@@ -1,212 +1,22 @@
+  </BaseLayout>
+
+  <div class="container">
+    <SchemaEdit :schema="null" ref="editor"></SchemaEdit>
```
Is this ref used somehow?

crazyscientist replied:
No, after defining it I noticed that I don't need it :joy:

commanderprice commented:
Value picker for filter for bool field looks a bit strange😅 
![image](https://user-images.githubusercontent.com/36228624/146389706-5adbcc63-0504-434c-a0ec-9d9faf52ffd8.png)

commanderprice commented on frontend/src/components/EditAttributes.vue:
```diff
@@ -0,0 +1,152 @@
+        </option>
+      </template>
+    </TextInput>
+    <Select label="Type" v-model="attributes[rowIndex].type" @change="onChange()"
+            :args="{ id: `initialattrType${rowIndex}`, disabled: !attr.isNew }"
+            :options="
+              Object.keys(ATTR_TYPES_NAMES).map((type) => {
+                return { value: type, text: ATTR_TYPES_NAMES[type] };
+              })"/>
```
I guess there's problem somewhere in here: when looking at schema details it displays types of all fields as `string` even if API returns other type

commanderprice commented:
Found one issue with schema updates: if you create new attribute, but don't save it and switch to another schema, you won't see it in list of attributes. However, update button will be active and it will actually add this attribute to switched schema😦

crazyscientist committed:
Fixed issues with filter dialog and changed icon/logo

commanderprice committed:
Commit some changes in traceability

commanderprice committed:
Rewrite traceability

Now traceability uses different, EAV-like data model which allows to keep track of different objects without adding new tables

crazyscientist committed:
Merge branch 'master' of github.com:commanderprice/aimaas into ui

crazyscientist committed:
Merge branch 'master' of github.com:commanderprice/aimaas into ui

 Conflicts:
	backend/traceability/schema.py

crazyscientist committed:
Refactored views for editing/creating entities

crazyscientist committed:
Convert API into a global application property

crazyscientist committed:
Merge branch 'master' of github.com:commanderprice/aimaas into ui

 Conflicts:
	backend/general_routes.py
	backend/tests/test_traceability_schema.py
	backend/traceability/entity.py
	backend/traceability/schema.py

crazyscientist committed:
Added new alert "plugin"

crazyscientist committed:
Converted alert into proper plugin

crazyscientist committed:
Merge branch 'master' of github.com:commanderprice/aimaas into ui

crazyscientist committed:
Merge branch 'master' of github.com:commanderprice/aimaas into ui

crazyscientist committed:
Merge branch 'master' of github.com:commanderprice/aimaas into ui

crazyscientist committed:
* Fixed issues with alert plugin/component
* Introduced new replacement component for changes

crazyscientist committed:
Refactored how changes are viewed/displayed

crazyscientist committed:
Merge branch 'master' of github.com:commanderprice/aimaas into ui

crazyscientist committed:
Fixed schema update

crazyscientist committed:
Merge branch 'master' of github.com:commanderprice/aimaas into ui

crazyscientist committed:
Updated change view for schemas

crazyscientist committed:
Fixed entity update

crazyscientist committed:
Redesign changes component

crazyscientist committed:
Merge branch 'master' of github.com:commanderprice/aimaas into ui

crazyscientist committed:
Redesign changes component

crazyscientist committed:
Merge branch 'master' of github.com:commanderprice/aimaas into ui

crazyscientist committed:
Added filter parameter for schema list and API function to delete schemas

crazyscientist committed:
Make sure alters are always in the viewport

crazyscientist committed:
Merge branch 'master' of github.com:commanderprice/aimaas into ui

crazyscientist committed:
Add indicator for required fields

crazyscientist committed:
Merge branch 'master' of github.com:commanderprice/aimaas into ui

crazyscientist committed:
Fix API issues

crazyscientist committed:
* Moved Pagination component
* Added missing filter params to EntityList

crazyscientist committed:
* Fixed issue with creating/updating entities
* Resolve referenced entities in list
* Fixed display issue in SearchPanel

crazyscientist committed:
* Removed attribute suggestions
* Fixed value binding in SelectInput

crazyscientist committed:
* Made API reactive
* Delete API method to fetch attributes
* Fixed issue with API login method
* Fixed issue with DateInput

crazyscientist committed:
Merge branch 'master' of github.com:commanderprice/aimaas into ui

crazyscientist committed:
Merge branch 'master' of github.com:commanderprice/aimaas into ui

crazyscientist committed:
Merge branch 'master' of github.com:commanderprice/aimaas into ui

crazyscientist committed:
Merge branch 'master' of github.com:commanderprice/aimaas into ui

crazyscientist committed:
* Pending Reviews
* Allow vertical confirmation dialog
* Fixed issue in ReferencedEntitySelect.vue when schema has multiple FKs
* Add `big` parameter to Placeholder.vue
* Show list of FKs correctly

crazyscientist committed:
Added capability to accept/decline reviews to UI

crazyscientist committed:
Merge branch 'master' of github.com:commanderprice/aimaas into ui

crazyscientist committed:
Finished review system

crazyscientist committed:
Added breadcrumbs to pending reviews page

crazyscientist committed:
Removed debug output

crazyscientist committed:
Fixed issue with unavailability of `randomUUID`

commanderprice commented:
I feel that pending reviews page doesn't provide enough details about them. For example, here we have an entity update request and entity create request, ~but you can't really tell the difference between them~ oh, okay, I noticed icons.
For update request you don't see what entity you are updating, and for create request it would be beneficial to see what schema we are dealing with.

Also, I noticed, that set of pending reviews doesn't change depending on schema, but maybe that's how you planned it to be

Another thing is that neither reviews count, nor the actual list itself doesn't update if you create a new review, you have to reload the page
![image](https://user-images.githubusercontent.com/36228624/147746010-4a3b4551-5c05-4a1d-96c9-1ba996cb5620.png)

commanderprice commented:
You'll probably want to think about how listed values are displayed😅

![image](https://user-images.githubusercontent.com/36228624/147746677-38889bb7-816e-485c-aa41-25a8b4b4a490.png)

Also when we display all columns (or not all, whatever) we can try to order by columns which are not orderable (these are lists, at least) which will lead to infinite load

commanderprice commented:
At entity edit page there is a line `Start typing to receive suggestions`. I believe it is for my reference list above. If it is so, then it has kind of misleading position😅 But more than that, there is nowhere to type to receive suggestions😥. Also it would be nice to have here links to chosen entities

![image](https://user-images.githubusercontent.com/36228624/147746923-28977b47-2161-4b02-9cfe-217396aa02fe.png)

commanderprice commented:
If you have a schema with field pointing to other schema and try to create entity pointing to entities of that schema, it *will* create this entity, but it won't be able to show these referenced entities. It looks like it tries to get those entities by id from *current* schema

commanderprice commented:
If I log in and reload page, I have to log in again 🤔 But that's probably not a big problem now, because user data is not used anywhere now

commanderprice commented:
Consider you have a schema with 5 pages of entities. You go to page 2, then select other schema

* If new schema has 2 or more pages, it will show you second page of entities
* If it doesn't, it will say that there's nothing to display

commanderprice commented:
Also you'll want to take care of horizontal overflow when there is a lot of pages that can't fit current screen width
![image](https://user-images.githubusercontent.com/36228624/147748136-0afda019-a997-414e-aff9-8c257575e3ba.png)

crazyscientist committed:
Reload pending reviews after updating/creating entity

crazyscientist commented:
> I feel that pending reviews page doesn't provide enough details about them.

I have added links to the schema and entity, if available.

> Also, I noticed, that set of pending reviews doesn't change depending on schema, but maybe that's how you planned it to be.

Yes, this is intentional. This view is supposed to show **all** pending requests.

> neither reviews count, nor the actual list itself doesn't update if you create a new review

Done, now it gets automatically updated. However, in the future we will need to worry about reloading this list, if another client submits change requests :wink:

crazyscientist committed:
Display list attributes nicely in entity list table

crazyscientist commented:
> You'll probably want to think about how listed values are displayed

Oops. Did that for FKs, but not the other types. :laughing: Done.

crazyscientist committed:
Fixes for referenced entities

crazyscientist commented:
> At entity edit page there is a line `Start typing to receive suggestions`

Oops, that does not describe the actual behavior... Changed it.

> Also it would be nice to have here links to chosen entities

No problem, done.

crazyscientist commented:
> If you have a schema with field pointing to other schema and try to create entity pointing to entities of that schema, it will create this entity, but it won't be able to show these referenced entities. It looks like it tries to get those entities by id from current schema

Yes, there was an issue. Should be fixed now...

crazyscientist committed:
Excluded list and FK attributes from ordering
